src/Controls/src/Core/Platform/Android/Extensions/FormattedStringExtensions.cs

```diff
@@ -131,49 +133,48 @@ public static SpannableString ToSpannableString(this Label label)
 			return spannable;
 		}
 
-		public static void RecalculateSpanPositions(this TextView textView, Label element, SpannableString spannableString, SizeRequest finalSize)
+		public static void RecalculateSpanPositions(this TextView textView, Label element, SpannableString spannableString, SizeRequest finalSize) =>
+			RecalculateSpanPositions(textView, element.Padding, element.FormattedText, spannableString);
+
+		internal static void RecalculateSpanPositions(this TextView textView, Thickness padding, FormattedString text) =>
+			RecalculateSpanPositions(textView, padding, text, textView.TextFormatted as SpannableStringInternal);
+
+		internal static void RecalculateSpanPositions(this TextView textView, Thickness padding, FormattedString text, SpannableStringInternal? spannableString)
 		{
-			if (element?.FormattedText?.Spans is null || element.FormattedText.Spans.Count == 0)
+			if (textView.Width <= 0 || textView.Height <= 0)
 				return;
 
-			var labelWidth = finalSize.Request.Width;
-			if (labelWidth <= 0 || finalSize.Request.Height <= 0)
+			if (spannableString is null || spannableString.IsDisposed())
 				return;
 
-			if (spannableString == null || spannableString.IsDisposed())
+			var strlen = spannableString.Length();
+			if (strlen == 0)
 				return;
 
 			var layout = textView.Layout;
-			if (layout == null)
+			if (layout is null)
 				return;
 
 			int next = 0;
 			int count = 0;
 
-			var padding = element.Padding;
 			var padLeft = (int)textView.Context.ToPixels(padding.Left);
 			var padTop = (int)textView.Context.ToPixels(padding.Top);
 
-#pragma warning disable CA1416
-			var strlen = spannableString.Length();
-#pragma warning restore CA1416
-
 			for (int i = 0; i < strlen; i = next)
 			{
-				var type = Java.Lang.Class.FromType(typeof(Java.Lang.Object));
-
-				var span = element.FormattedText.Spans[count];
+				var span = text.Spans[count];
 
 				count++;
 
 				if (string.IsNullOrEmpty(span.Text))
 					continue;
 
 				// Find the next span
-				next = spannableString.NextSpanTransition(i, spannableString.Length(), type);
+				next = spannableString.NextSpanTransition(i, spannableString.Length(), JavaLangObjectClass);
```

Comment:
I think you can skip the JNI and the Java side reflection entirely by passing null https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/SpannableStringInternal.java#428

https://developer.android.com/reference/android/text/Spanned#nextSpanTransition(int,%20int,%20java.lang.Class)

Reply:
I saw that and gave it a go, but I think on some older android it threw an exception. I can check again and see what happened... It may have been something else.

Reply:
That [check](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/lollipop-release/core/java/android/text/SpannableStringInternal.java#292) is there as far back as Lolipop (API 21) which the lowest that .NET supports.

Reply:
Thanks! I see that null can be used here and NOT with the one below - even if the API says it can. https://github.com/xamarin/xamarin-android/issues/8167